docs/src/main/asciidoc/async-message-passing.adoc

```diff
@@ -0,0 +1,290 @@
+
+[source, java]
+----
+package org.acme.vertx;
+
+import io.quarkus.vertx.ConsumeEvent;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class GreetingService {
+
+    @ConsumeEvent                           // <1>
+    public String consume(String name) {    // <2>
+        return name.toUpperCase();
+    }
+}
+----
+<1> If not set, the address is the fully qualified name of the bean, for instance, in this snippet it's `org.acme.NameService`
+<2> The method parameter is the message body. If the method returns _something_ it's the message response.
+
+// TODO Revisit this once the blocking PR is integrated - https://github.com/quarkusio/quarkus/pull/1646
+IMPORTANT: the code consuming the event must be _non-blocking_, as it's called on the Vert.x event loop.
+
+=== Configuring the address
+
+The `@ConsumeEvent` annotation can be configured to set the address:
+
+[source, java]
+----
+@ConsumeEvent("greeting")               // <1>
+public String consume(String name) {
+    return name.toUpperCase();
+}
+----
+<1> Receive the messages sent on the `greeting` address
+
+=== Consuming without replying
+
+In the previous examples, the consumer returns a result, but it's not necessary:
+
+[source,java]
+----
+@ConsumeEvent("greeting")
+public void consume(String event) {
+    // Do something with the event
+}
+----
+
+=== Replying
+
+The _return_ of a method annotated with `@ConsumeEvent` is used as response to the incoming message.
+For instance, in the following snippet, the returned `String` is the response.
+
+[source, java]
+----
+@ConsumeEvent("greeting")
+public String consume(String name) {
+    return name.toUpperCase();
+}
+----
+
+You can also return a `CompletionStage<T>` to handle asynchronous reply:
+
+[source, java]
+----
+@ConsumeEvent("greeting")
+public CompletionStage<String> consume2(String name) {
+    return CompletableFuture.supplyAsync(name::toUpperCase);
+}
+----
+
+=== Dealing with messages
+
+As said above, this mechanism is based on the Vert.x event bus. So, you can also use `Message` directly:
+
+[source, java]
+----
+@ConsumeEvent("greeting")
+public void consume(Message<String> msg) {
+    System.out.println(msg.address());
+    System.out.println(msg.body());
+}
+----
+
+== Sending messages
+
+Ok, we have seen how to receive messages, let's now switch to the _other side_: the sender.
+Sending and publishing messages use the Vert.x event bus:
+
+[source, java]
+----
+package org.acme;
+
+import io.vertx.axle.core.eventbus.EventBus;
+import io.vertx.axle.core.eventbus.Message;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import java.util.concurrent.CompletionStage;
+
+@Path("/async")
+public class EventResource {
+
+  @Inject
+  EventBus bus;                                                        // <1>
+
+  @GET
+  @Path("/{name}")
+  public CompletionStage<String> hello(@PathParam("name") String name) {
```

Comment:
RESTEasy has a version of `@PathParam` that does not require duplicating the parameter name. Probably we want to use this.

Reply:
Didn't know this!